pkg/liqo-controller-manager/namespaceOffloading-controller/namespaceoffloading_controller.go

```diff
@@ -0,0 +1,214 @@
+// --> NamespaceOffloading.Status.RemoteNamespaceName.
+// --> NamespaceOffloadingController finalizer.
+// --> NamespaceMap.Spec.DesiredMapping, only for my namespace entries.
+
+// Reconcile requires creation of remote namespaces according to ClusterSelector field.
+func (r *NamespaceOffloadingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	namespaceOffloading := &offv1alpha1.NamespaceOffloading{}
+	if err := r.Get(context.TODO(), types.NamespacedName{
+		Namespace: req.Namespace,
+		Name:      liqoconst.DefaultNamespaceOffloadingName,
+	}, namespaceOffloading); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Infof("There is no namespaceOffloading in namespace '%s'", req.Namespace)
+			return ctrl.Result{}, nil
+		}
+		klog.Errorf("%s --> Unable to get namespaceOffloading for the namespace '%s'", err, req.Namespace)
+		return ctrl.Result{}, err
+	}
+
+	// Get all NamespaceMaps in the cluster and create a Map 'cluster-id : *NamespaceMap'
+	namespaceMapsList := &mapsv1alpha1.NamespaceMapList{}
+	clusterIDMap := make(map[string]*mapsv1alpha1.NamespaceMap)
+	if err := r.getClusterIDMap(ctx, namespaceMapsList, clusterIDMap); err != nil {
+		return ctrl.Result{}, err
+	}
+	// There are no NamespaceMap in the cluster
+	if len(clusterIDMap) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// If deletion timestamp is set, it starts deletion logic and waits until all remote Namespaces
+	// associated with this resource are deleted.
+	if !namespaceOffloading.GetDeletionTimestamp().IsZero() {
+		if err := r.deletionLogic(ctx, namespaceOffloading, clusterIDMap); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	// Initialize NamespaceOffloading Resource if it has been just created.
+	if !ctrlutils.ContainsFinalizer(namespaceOffloading, namespaceOffloadingControllerFinalizer) {
+		if err := r.initialConfiguration(ctx, namespaceOffloading); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	// Request creation of remote Namespaces in according with ClusterSelector field.
+	if err := r.enforceClusterSelector(namespaceOffloading, clusterIDMap); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If there is at least one remote Namespace add liqo scheduling label to the Namespace.
+	if len(clusterIDMap) != len(namespaceMapsList.Items) {
+		// this label will trig liqo webhook
+		if err := addLiqoSchedulingLabel(ctx, r.Client, namespaceOffloading.Namespace); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// Todo: how to awake this controller for every NamespaceOffloading when a new NamespaceMap is created (or recreated).
+func namespaceOffloadingPredicate() predicate.Predicate {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if !(e.ObjectNew.GetDeletionTimestamp().IsZero()) && slice.ContainsString(e.ObjectNew.GetFinalizers(),
+				namespaceOffloadingControllerFinalizer, nil) {
+				return true
+			}
+			return false
```

Comment:
1. Thank you
2. In the deletion, this check is made by means of the finalizer, only a resource with the right name should have the NamespaceOffloading finalizer, do you suggest adding it as a strong check?
3. Exactly

Reply:
2. I would be a bit more verbose (adding the explicit check for the name) to be on the safe side, although it should not make any effective difference. Just a matter of preference
3. I would add a brief comment before the predicate about this expected behavior, otherwise it may look strange at first sight